Deploy ISPN + KC + AuroraDB in cross-site configuration

pruivo opened the pull request.

ryanemerson commented on provision/aws/rds/aurora_create_peering_connection.sh:
```diff
@@ -63,10 +65,24 @@ ROSA_PUBLIC_ROUTE_TABLE_ID=$(aws ec2 describe-route-tables \
+  --output text)
+
+if [ "${EXISTS}" == "" ]; then
+  aws ec2 delete-route \
```
Why do we need to delete and then create if we've already checked that the route doesn't exist?

pruivo replied:
I had some runs where `EXISTS` was empty but the route really existed. I'll remove it since I'm pretty sure I was doing something wrong back then :rofl:

ahus1 commented:
@pruivo / @ryanemerson - I rebased this PR and resolved a conflict in `infinispan.yaml` ... it seems like some of the changes were already part of the other PR #469.

mhajas commented on provision/rosa-cross-dc/Taskfile.yml:
```diff
@@ -0,0 +1,308 @@
+    preconditions:
+      - test -f {{.ISPN_DIR}}/.task/kubecfg/{{.ROSA_CLUSTER_NAME}}
+
+  install-keycloak:
```
Can we move this to `provision/minikube/Utils.yaml` and reuse it in both `minikube/Taskfile` and `openshift/Taskfile`? It does not need to be done in this PR, just curious if you thought about this and whether there are some limitations. I can try to do this tomorrow.

pruivo replied:
I tried to move the `keycloak` task from `openshift/Taskfile.yaml` to another taskfile to be included by both, but `isup.sh` was giving me an issue because it was reading the `.env` file from `openshift/.env` and failing. I'm not sure if it is worth the effort since it requires some changes; at the end of the day, it depends on how often the helm chart changes.

pruivo commented:
> @pruivo / @ryanemerson - I rebased this PR and resolved a conflict in `infinispan.yaml` ... it seems like some of the changes were already part of the other PR #469.

thanks @ahus1. Yes, some changes were already present in the other PR since I was hitting the issue while testing this PR.

pruivo commented on provision/aws/rds/aurora_create_peering_connection.sh:
```diff
@@ -61,10 +63,24 @@ ROSA_PUBLIC_ROUTE_TABLE_ID=$(aws ec2 describe-route-tables \
+    --route-table-id ${ROSA_PUBLIC_ROUTE_TABLE_ID} \
+    --destination-cidr-block ${AURORA_VPC_CIDR} \
+    --vpc-peering-connection-id ${PEERING_CONNECTION_ID} || true
+fi
```
I can do that!

pruivo commented:
Rebased and tested using the changes from @ryanemerson. Tested with 2 ROSA clusters, one in `eu-west-1` and the other in `eu-west-3` and Aurora DB in `eu-west-2`. No connection errors were observed. 

Added a README for completeness.

ryanemerson commented on provision/rosa-cross-dc/Taskfile.yml:
```diff
@@ -0,0 +1,314 @@
+        - ROSA_CLUSTER_NAME
+        - AURORA_CLUSTER
+        - AURORA_REGION
+    cmd: CLUSTER_NAME="{{.ROSA_CLUSTER_NAME}}" KUBECONFIG="{{.ISPN_DIR}}/.task/kubecfg/{{.ROSA_CLUSTER_NAME}}" SKIP_ROSA_LOGIN=true {{.RDS_DIR}}/aurora_create_peering_connection.sh
```
Can we make use of `env:` here instead of passing as part of the command?

This also applies to many of the other tasks in this file.

pruivo replied:
The semantics are weird when you have some variables in your environment (example, using `export`). You can give it a try:
test.sh
```bash
#!/bin/bash

echo "Message=${MESSAGE}"
```
Taskfile.yaml
```yaml
version: "3"
tasks:
  default:
    - task: print
      vars:
        MESSAGE: "print this"
    - task: print2
      vars:
        MESSAGE: "print this"
  print:
    env:
      MESSAGE: "{{.MESSAGE}}"
    cmds:
      - ./test.sh
  print2:
    cmds:
      - MESSAGE="{{.MESSAGE}}" ./test.sh
```
As you see below, it prefer to uses the value from `export` instead of using the value defined in the yaml file. It caused a lot of headaches so I decided to play safe (and ugly).
```bash
$ export MESSAGE="not this"
$ go-task 
task: [print] ./test.sh
Message=not this
task: [print2] MESSAGE="print this" ./test.sh
Message=print this
```

ryanemerson replied:
Ok, thanks for explaining :+1:

ryanemerson reviewed (commented):
An initial deployment failed for me due to an API timeout (I don't *think* it was an issue with your scripts), but now when I try to do `task undeploy` I get the following:

```
task: Task "common:env" is up to date
cat: .task/ocp-prefix-gh-ryan-e1: No such file or directory
task: Failed to run task "undeploy": task: Command "cat .task/ocp-prefix-gh-ryan-e1" failed: exit status 1
```

ryanemerson commented:
Using the below command, the Infinispan deployment never comes ready due to an authorization issue.

```
AURORA_CLUSTER=ryan-test AURORA_REGION=us-east-2 ROSA_CLUSTER_NAME_1=gh-ryan-n1 ROSA_CLUSTER_NAME_2=gh-ryan-e1 task
```

Relevant part of the operator logs on `gh-ryan-n1`:

```
2023-08-11T11:14:52.361Z	ERROR	controllers.Infinispan	could not connect to remote location URL	{"infinispan": "remerson-keycloak/infinispan", "URL": "openshift://api.gh-ryan-n1.wasw.p1.openshiftapps.com:6443", "error": "Unauthorized"}
github.com/go-logr/zapr.(*zapLogger).Error
	/opt/app-root/src/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan/handler/configure.appendRemoteLocation
	/workspace/pkg/reconcile/pipeline/infinispan/handler/configure/xsite.go:106
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan/handler/configure.searchRemoteSites
	/workspace/pkg/reconcile/pipeline/infinispan/handler/configure/xsite.go:89
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan/handler/configure.XSite
	/workspace/pkg/reconcile/pipeline/infinispan/handler/configure/xsite.go:52
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan.HandlerFunc.Handle
	/workspace/pkg/reconcile/pipeline/infinispan/api.go:40
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan/pipeline.invokeHandler
	/workspace/pkg/reconcile/pipeline/infinispan/pipeline/pipeline.go:60
github.com/infinispan/infinispan-operator/pkg/reconcile/pipeline/infinispan/pipeline.(*impl).Process
	/workspace/pkg/reconcile/pipeline/infinispan/pipeline/pipeline.go:42
github.com/infinispan/infinispan-operator/controllers.(*InfinispanReconciler).Reconcile
	/workspace/controllers/infinispan.go:238
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
	/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.19.4/pkg/util/wait/wait.go:99
2023-08-11T11:14:52.361Z	INFO	controllers.Infinispan	Done	{"infinispan": "remerson-keycloak/infinispan", "requeue": true, "requeueAfter": "0s", "error": "unable to search remote sites: Unauthorized"}

```

pruivo commented on provision/rosa-cross-dc/Taskfile.yml:
```diff
@@ -0,0 +1,314 @@
+          CROSS_DC_HOT_ROD_PASSWORD: "{{.RS_HOT_ROD_PASSWORD}}"
+
+
+  fetch-ocp-hostname-prefix:
```
this task will become redundant after https://github.com/keycloak/keycloak-benchmark/pull/484/files#diff-246cc39ce6408ddc2fe1cf751d9d66f9d87c95fc82ea8969990490e51ba19625R24 is merged.

pruivo commented:
@ryanemerson I think your error is fixed by https://github.com/keycloak/keycloak-benchmark/pull/484

pruivo commented:
rebased. Hopefully, nothing is broken :crossed_fingers:

ryanemerson reviewed (approved):
I just tested again and it's working as expected now :slightly_smiling_face:
